### PR TITLE
Make api of visibility and permission extensions more robust

### PR DESCRIPTION
**Pull request by sliverc**
Instead of relying on naming rather use decorator with referencing import directly. This way typos can be noticed during import.

### CODE REVIEW

File: caluma/mutation.py [hunk 1]
```diff
@@ -125,7 +128,7 @@ def __init_subclass_with_meta__(
         _meta.return_field_type = return_field_type
 
         input_fields = yank_fields_from_attrs(input_fields, _as=InputField)
-        super(SerializerMutation, cls).__init_subclass_with_meta__(
+        super(Mutation, cls).__init_subclass_with_meta__(
```
Comment (winged) [inline caluma/mutation.py]: Is there a reason why we're using python2-style `super()` here?

File: caluma/permissions.py [hunk 2]
```diff
@@ -1,36 +1,87 @@
...
 
-    def has_object_permission(self, mutation, info, instance):
-        """
-        Check object permission of specific mutation or otherwise return True.
+        object_perm_fns = inspect.getmembers(
+            self, lambda m: hasattr(m, "_object_permission")
+        )
+        self._object_permissions = {
+            fn._object_permission: fn for _, fn in object_perm_fns
+        }
```
Comment (winged) [inline caluma/permissions.py]: It'd probably be a good idea to safeguard against duplicate permission methods here. Throw an exception if two methods are tagged for the same permission class

File: caluma/visibilities.py [hunk 3]
```diff
@@ -1,31 +1,54 @@
...
 
-        Checks for node specific implement by its snake name.
 
-        e.g. for Node WorkItem it will call `get_queryset_for_work_item` if available.
-        """
-        queryset = self.get_base_queryset(node, queryset, info)
+class BaseVisibility(object):
+    """Basic visibility classes to be extended by any visibility implementation.
+
+    In combination with the decorator `@queryset_for` a custom visibility class
```
Comment (winged) [inline caluma/visibilities.py]: Naming - the decorator is called `filter_queryset_for`, not `queryset_for`

File: caluma/visibilities.py [hunk 4]
```diff
@@ -1,31 +1,54 @@
...
+        self._filter_querysets_for = {
+            fn._filter_queryset_for: fn for _, fn in queryset_fns
+        }
 
-        method_name = f"get_queryset_for_{to_snake_case(node.__name__)}"
-        get_queryset_for_node = getattr(self, method_name, None)
-        if get_queryset_for_node is not None:
-            queryset = get_queryset_for_node(node, queryset, info)
+    def get_queryset(self, node, queryset, info):
+        for cls in reversed(node.mro()):
```
Comment (winged) [inline caluma/visibilities.py]: Just out of curiosity: Why are you iterating in reversed MRO here? I think filtering is "additive", so shouldn't really matter

File: caluma/mutation.py (hunk 1 above)
Comment (sliverc) [reply caluma/mutation.py]: Changed

File: caluma/permissions.py (hunk 2 above)
Comment (sliverc) [reply caluma/permissions.py]: Added safeguarding

File: caluma/visibilities.py (hunk 3 above)
Comment (sliverc) [reply caluma/visibilities.py]: well spotted. Changed.

File: caluma/visibilities.py (hunk 4 above)
Comment (sliverc) [reply caluma/visibilities.py]: Removed it as it makes no difference and is just more code... :)

**Comment by sliverc** [comment]
@winged 
Changes for loading `permission_classes` and `visibility_classes` only once app is ready needs a lot of files being moved around.
Hence I will create a single PR for this and this PR can be reviewed/merged before.

**Pull request merged**